src/Microsoft.IdentityModel.Tokens/Validation/Results/Details/AudienceValidationError.cs

```diff
@@ -10,31 +10,43 @@ namespace Microsoft.IdentityModel.Tokens
 {
     internal class AudienceValidationError : ValidationError
     {
-        private IList<string>? _invalidAudiences;
+        private IList<string>? _tokenAudiences;
+        private IList<string>? _validAudiences;
+
+        // stack frames associated with AudienceValidationErrors
+        internal static StackFrame? ValidationParametersNull;
+        internal static StackFrame? AudiencesNull;
+        internal static StackFrame? AudiencesCountZero;
+        internal static StackFrame? ValidationParametersAudiencesCountZero;
+        internal static StackFrame? ValidateAudienceFailed;
 
         public AudienceValidationError(
             MessageDetail messageDetail,
+            ValidationFailureType failureType,
             Type exceptionType,
             StackFrame stackFrame,
-            IList<string>? invalidAudiences)
-            : base(messageDetail, ValidationFailureType.AudienceValidationFailed, exceptionType, stackFrame)
+            IList<string>? tokenAudiences,
+            IList<string>? validAudiences)
+            : base(messageDetail, failureType, exceptionType, stackFrame)
         {
-            _invalidAudiences = invalidAudiences;
-        }
-
-        internal override void AddAdditionalInformation(ISecurityTokenException exception)
-        {
-            if (exception is SecurityTokenInvalidAudienceException invalidAudienceException)
-                invalidAudienceException.InvalidAudience = Utility.SerializeAsSingleCommaDelimitedString(_invalidAudiences);
+            _tokenAudiences = tokenAudiences;
+            _validAudiences = validAudiences;
         }
 
         /// <summary>
         /// Creates an instance of an <see cref="Exception"/> using <see cref="ValidationError"/>
         /// </summary>
         /// <returns>An instance of an Exception.</returns>
-        public override Exception GetException()
+        internal override Exception GetException()
         {
-            return new SecurityTokenInvalidAudienceException(MessageDetail.Message) { InvalidAudience = Utility.SerializeAsSingleCommaDelimitedString(_invalidAudiences) };
+            if (ExceptionType == typeof(SecurityTokenArgumentNullException))
+            {
+                return new SecurityTokenArgumentNullException(MessageDetail.Message);
```

Comment:
This is a common case, would be best handled in the base method instead of repeating it in every `ValidationError` subclass.